pkg/pipeline/BuildPipelineConfigService.go

```diff
@@ -1264,6 +1272,319 @@ func (impl *CiPipelineConfigServiceImpl) UpdateCiTemplate(updateRequest *bean.Ci
+			return nil, errors.New("invalid request payload, both switchFromCiPipelineId and switchFromExternalCiPipelineId cannot be set in the payload")
+		}
+
+		//get the ciPipeline
+		var switchFromType bean.PipelineType
+		var ciPipeline *pipelineConfig.CiPipeline
+		var err error
+		if request.SwitchFromExternalCiPipelineId != 0 {
+			switchFromType = bean.EXTERNAL
+		} else {
+			ciPipeline, err = impl.ciPipelineRepository.FindById(request.SwitchFromCiPipelineId)
+			if err != nil {
+				impl.logger.Errorw("error in finding ci-pipeline by id", "ciPipelineId", request.SwitchFromCiPipelineId, "err", err)
+				return nil, err
+			}
+		}
+
+		//validate request if it's a switch request
+		err = impl.validateCiPipelineSwitch(request.SwitchFromCiPipelineId, request.SwitchFromExternalCiPipelineId, request.CiPipeline.PipelineType, switchFromType)
+		if err != nil {
+			impl.logger.Errorw("error occurred when validating ci-pipeline switch", "err", err)
+			return nil, err
+		}
+
+		//delete old pipeline and it's appworkflow mapping
+		tx, err = impl.ciPipelineRepository.StartTx()
+		if err != nil {
+			impl.logger.Errorw("error in starting transaction", "err", err)
+			return nil, err
+		}
+		oldAppWorkflowMapping, err = impl.deleteOldPipelineAndWorkflowMappingBeforeSwitch(tx, ciPipeline, request.SwitchFromExternalCiPipelineId, request.UserId)
+		if err != nil {
+			impl.logger.Errorw("error in deleting old ci-pipeline and getting the appWorkflow mapping of that", "err", err, "userId", request.UserId)
+			return nil, err
+		}
+	}
+
+	impl.logger.Debugw("create patch request")
+	ciConfig.CiPipelines = []*bean.CiPipeline{request.CiPipeline} //request.CiPipeline
+
+	pipelineExists, err := impl.ciPipelineRepository.CheckIfPipelineExistsByNameAndAppId(request.CiPipeline.Name, request.AppId)
+	if err != nil && err != pg.ErrNoRows {
+		impl.logger.Errorw("error in fetching pipeline by name, FindByName", "err", err, "patch cipipeline name", request.CiPipeline.Name)
+		return nil, err
+	}
+
+	if pipelineExists {
+		impl.logger.Errorw("pipeline name already exist", "err", err, "patch cipipeline name", request.CiPipeline.Name)
+		return nil, fmt.Errorf("pipeline name already exist")
+	}
+
+	res, err := impl.addpipelineToTemplate(ciConfig)
+	if err != nil {
+		impl.logger.Errorw("error in adding pipeline to template", "ciConf", ciConfig, "err", err)
+		return nil, err
+	}
+
+	//go and update all the app workflow mappings of old ci_pipeline with new ci_pipeline_id.
+	if request.IsSwitchCiPipelineRequest() {
+		err = impl.updateLinkedAppWorkflowMappings(tx, oldAppWorkflowMapping, res.AppWorkflowMapping)
+		if err != nil {
+			impl.logger.Errorw("error in updating app workflow mappings", "err", err)
+			return nil, err
+		}
+		//we don't store ci-pipeline-id in pipeline table for
+		if request.SwitchFromCiPipelineId > 0 {
+			// ciPipeline id is being set in res object in the addpipelineToTemplate method.
+			//
```

Comment:
would be cleaner if we move this entire logic to a function and do something like this

```go

	pipelineExists, err := impl.ciPipelineRepository.CheckIfPipelineExistsByNameAndAppId(request.CiPipeline.Name, request.AppId)
	if err != nil && err != pg.ErrNoRows {
		impl.logger.Errorw("error in fetching pipeline by name, FindByName", "err", err, "patch cipipeline name", request.CiPipeline.Name)
		return nil, err
	}

	if pipelineExists {
		impl.logger.Errorw("pipeline name already exist", "err", err, "patch cipipeline name", request.CiPipeline.Name)
		return nil, fmt.Errorf("pipeline name already exist")
	}
	
	if  request.IsSwitchCiPipelineRequest() {
	handleSwitchRequest () // you can again call impl.addpipelineToTemplate(ciConfig) inside
	} else {
res, err := impl.addpipelineToTemplate(ciConfig)
}

Reply:
done